File: src/services/bedrock.js
import { BedrockAgentRuntimeClient, InvokeAgentCommand } from '@aws-sdk/client-bedrock-agent-runtime';
import { BedrockAgentClient, ListIngestionJobsCommand, StartIngestionJobCommand, GetKnowledgeBaseCommand, GetDataSourceCommand, GetAgentCommand, ListDataSourcesCommand, GetIngestionJobCommand } from "@aws-sdk/client-bedrock-agent";
import { formatDate } from '../utils/date.js';
/*
Purpose and Operations:

BedrockAgentClient: Used for management operations like creating, updating, deleting, and configuring agents and knowledge bases. This includes operations like CreateAgent, DeleteKnowledgeBase, GetKnowledgeBase, etc.
BedrockAgentRuntimeClient: Used for runtime interactions with existing agents and knowledge bases. This includes operations like InvokeAgent, Retrieve, RetrieveAndGenerate, etc.
*/

// Formats the traceback information in a Slack-friendly way
function formatTraceback(traceback) {
  if (!traceback) return "No traceback information available";

  let formattedOutput = "```\n🤖 Agent Traceback\n================\n";

  // Format each step in the traceback
  if (traceback.steps && Array.isArray(traceback.steps)) {
    traceback.steps.forEach((step, index) => {
      formattedOutput += `\n🔹 Step ${index + 1}: ${step.type}\n`;
      formattedOutput += `📊 Status: ${step.status}\n`;

      if (step.input) {
        formattedOutput += "📥 Input:\n";
        formattedOutput += JSON.stringify(step.input, null, 2) + "\n";
      }

      if (step.output) {
        formattedOutput += "📤 Output:\n";
        formattedOutput += JSON.stringify(step.output, null, 2) + "\n";
      }

      if (step.error) {
        formattedOutput += "❌ Error:\n";
        formattedOutput += JSON.stringify(step.error, null, 2) + "\n";
      }

      formattedOutput += "----------------------------------------\n";
    });
  }

  formattedOutput += "```";
  return formattedOutput;
}

// Invoke the Bedrock agent with the provided input text
async function invokeBedrockAgent({inputText, sessionId, attachments = [], includeTraceback = false}) {
  console.log(`Session Sample ID: ${sessionId}`);
  try {
    // Create AWS SDK client with credentials and region
    const client = new BedrockAgentRuntimeClient({
      region: process.env.AWS_BEDROCK_REGION,
      logger: console,
    });

    // Set up the parameters for the InvokeAgent operation
    const input = {
      agentAliasId: process.env.AWS_BEDROCK_AGENT_ALIAS_ID,
      agentId: process.env.AWS_BEDROCK_AGENT_ID,
      sessionId,
      enableTrace: true,
      inputText
    };

    // Add attachments if provided
    if (attachments?.length > 0) {
      input.files = [];
      attachments.forEach(attachment => {
        input.files.push({
          name: attachment.name,
          source: {
            byteContent: {
              data: Buffer.from(attachment.data).toString('base64'),
              mediaType: attachment.mediaType
            },
            sourceType: 'BYTE_CONTENT'
          },
          useCase: 'CHAT'
        });
      });
    }

    // Create and execute the InvokeAgent command
    const command = new InvokeAgentCommand(input);
    const output = await client.send(command);

    // Process the streaming response
    let result = '';
    let traceback = null;
    console.log(`length of output events:`, output.completion);

    for await (const event of output.completion) {
      if (event.chunk) {
        result += new TextDecoder().decode(event.chunk.bytes);
      }
      // Capture traceback information if available
      if (event.trace) {
        traceback = event.trace;
      }
    }

    console.log('AWS Bedrock agent response:', result);

    // Return both response and formatted traceback if requested
    if (includeTraceback) {
      return {
        response: result,
        traceback: formatTraceback(traceback)
      };
    }

    return result;

  } catch (error) {
    console.error('Error invoking Bedrock agent:', error);
    throw error;
  }
}

// Retrieves the current status and configuration of the knowledge base
async function getKnowledgeBaseStatus() {
  const client = new BedrockAgentClient({ region: process.env.AWS_BEDROCK_REGION });

  try {
    const command = new GetKnowledgeBaseCommand({
      knowledgeBaseId: process.env.AWS_BEDROCK_KNOWLEDGE_BASE_ID
    });

    const response = await client.send(command);
    console.log("Knowledge base response:", response);
    const formattedResponse = `Knowledge Base: ${response.knowledgeBase.name}\nStatus: ${response.knowledgeBase.status}\nCreated At: ${formatDate(response.knowledgeBase.createdAt)}\nUpdated At: ${formatDate(response.knowledgeBase.updatedAt)}`;
    return formattedResponse;
  } catch (error) {
    console.error("Error fetching knowledge base status:", error);
    throw error;
  }
}

// Retrieves the current status and configuration of the agent
async function getAgentStatus() {
  const client = new BedrockAgentClient({ region: process.env.AWS_BEDROCK_REGION });

  try {
    const command = new GetAgentCommand({
      agentId: process.env.AWS_BEDROCK_AGENT_ID
    });

    const response = await client.send(command);
    console.log("Agent response:", response);
    const formattedResponse = `Agent Name: ${response.agent.agentName}\nAgent ID: ${response.agent.agentId}\nStatus: ${response.agent.agentStatus}\nFoundation Model: ${response.agent.foundationModel}\nCreated At: ${formatDate(response.agent.createdAt)}\nUpdated At: ${formatDate(response.agent.updatedAt)}`;
    return formattedResponse;
  } catch (error) {
    console.error("Error fetching agent status:", error);
    throw error;
  }
}


// Retrieves the metadata for the data source
async function getDataSource() {
  const client = new BedrockAgentClient({ region: process.env.AWS_BEDROCK_REGION });

  // Retireve the last ingestion job for the data source
  // TODO: Update to loop through all data sources to grab the last ingestion job
  const command = new ListIngestionJobsCommand({
    knowledgeBaseId: process.env.AWS_BEDROCK_KNOWLEDGE_BASE_ID,
    dataSourceId: process.env.AWS_BEDROCK_DATA_SOURCE_ID,
    maxResults: 1,
    sortBy: {
      attribute: "STARTED_AT",
      order: "DESCENDING",
    }
  });

  try {
    let formattedLines = '';
    const response = await client.send(command);

    console.log("Last synchronized:", response);

    //Check if we have ingestion jobs
    if (response.ingestionJobSummaries && response.ingestionJobSummaries.length > 0) {
      // Display information for each job
      response.ingestionJobSummaries.forEach(ingestionJob => {
        console.log(`Knowledge Base ID: ${ingestionJob.knowledgeBaseId}`);
        console.log(`Data Source ID: ${ingestionJob.dataSourceId}`);
        console.log(`Status: ${ingestionJob.status}`);
        console.log(`Started At: ${ingestionJob.startedAt}`);
        console.log(`Last Sync Time: ${ingestionJob.updatedAt}`);
        console.log("--------------------------");

        const formattedDate = formatDate(ingestionJob.updatedAt);
        formattedLines += `Data Source: ${ingestionJob.dataSourceId}\nKnowledge Base: ${ingestionJob.knowledgeBaseId}\nMessage: ${ingestionJob.description}\nStatus: ${ingestionJob.status}\nLast sync: ${formattedDate}\n\n`;
      });
    } else {
      console.log("No data sources found for this knowledge base.");
    }

    return formattedLines;
  } catch (error) {
    console.error("Error fetching knowledge base info:", error);
  }
}

// Lists all data sources associated with the knowledge base
async function listDataSources() {
  const client = new BedrockAgentClient({ region: process.env.AWS_BEDROCK_REGION });

  try {
    const command = new ListDataSourcesCommand({
      knowledgeBaseId: process.env.AWS_BEDROCK_KNOWLEDGE_BASE_ID
    });

    const response = await client.send(command);
    console.log("Data source list response:", response);
    const formattedResponse = `${response.dataSourceSummaries.map(source => `Data Source: ${source.dataSourceId}\n Name: ${source.name}\n Status: ${source.status}\n Updated At: ${formatDate(source.updatedAt)}\n\n`).join('')}`;
    return formattedResponse;
  } catch (error) {
    console.error("Error listing data sources:", error);
    throw error;
  }
}

// Trigger a manual sync of the data source
// TODO Set to admins or team leads only to prevent too many triggers
async function syncDataSource() {
  const client = new BedrockAgentClient({ region: process.env.AWS_BEDROCK_REGION });

  const command = new StartIngestionJobCommand({
    knowledgeBaseId: process.env.AWS_BEDROCK_KNOWLEDGE_BASE_ID,
    dataSourceId: process.env.AWS_BEDROCK_DATA_SOURCE_ID,
    description: "Manual sync triggered on " + new Date().toISOString()
  });

  try {
    const response = await client.send(command);
    console.log("Sync job started successfully:", response);
    const formattedResponse = `Data Source Sync Initiated. Data Source: ${response.ingestionJob.dataSourceId}\nKnowledge Base: ${response.ingestionJob.knowledgeBaseId}\nJob ID: ${response.ingestionJob.ingestionJobId}\nStatus: ${response.ingestionJob.status}`;
    return formattedResponse;
  } catch (error) {
    console.error("Error starting knowledge base sync:", error);
    throw error;
  }
}

// Retrieves the configuration details of a specific data source
async function getDataSourceConfig() {
  const client = new BedrockAgentClient({ region: process.env.AWS_BEDROCK_REGION });

  try {
    const command = new GetDataSourceCommand({
      knowledgeBaseId: process.env.AWS_BEDROCK_KNOWLEDGE_BASE_ID,
      dataSourceId: process.env.AWS_BEDROCK_DATA_SOURCE_ID
    });

    const response = await client.send(command);
    console.log("Data source response:", response);
    const formattedResponse = `Data Source: ${response.dataSource.name}\nStatus: ${response.dataSource.status}\nConfiguration: Type: ${response.dataSource.dataSourceConfiguration.type}\nCreated At: ${formatDate(response.dataSource.createdAt)}\nUpdated At: ${formatDate(response.dataSource.updatedAt)}`;
    return formattedResponse;
  } catch (error) {
    console.error("Error fetching data source configuration:", error);
    throw error;
  }
}

// Retrieves the status and details of a specific ingestion job
async function getIngestionJobStatus(jobId) {
  const client = new BedrockAgentClient({ region: process.env.AWS_BEDROCK_REGION });

  try {
    const command = new GetIngestionJobCommand({
      knowledgeBaseId: process.env.AWS_BEDROCK_KNOWLEDGE_BASE_ID,
      dataSourceId: process.env.AWS_BEDROCK_DATA_SOURCE_ID,
      ingestionJobId: jobId
    });

    const response = await client.send(command);
    console.log("Ingestion job response:", response);
    const formattedResponse = `Ingestion Job: ${response.ingestionJob.ingestionJobId}\nStatus: ${response.ingestionJob.status}\nStarted At: ${formatDate(response.ingestionJob.startedAt)}\nUpdated At: ${formatDate(response.ingestionJob.updatedAt)}\nStatistics: ${response.ingestionJob.statistics}\nFailure Reasons: ${response.ingestionJob.failureReasons.length > 0 ? response.ingestionJob.failureReasons.join('\n') : 'None'}`;
    return formattedResponse;
  } catch (error) {
    console.error("Error fetching ingestion job status:", error);
    throw error;
  }
}

export {
  invokeBedrockAgent,
  getDataSource,
  syncDataSource,
  getKnowledgeBaseStatus,
  getDataSourceConfig,
  getAgentStatus,
  listDataSources,
  getIngestionJobStatus
};

File: src/services/bedrock.go
package services

import (
	"context"
	"fmt"
	"os"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	bedrockagent "github.com/aws/aws-sdk-go-v2/service/bedrockagent"
	bedrockagentruntime "github.com/aws/aws-sdk-go-v2/service/bedrockagentruntime"
	bedrockagentruntime_types "github.com/aws/aws-sdk-go-v2/service/bedrockagentruntime/types"

	"slack-rag-server/src/types"
)

// BedrockService provides methods for interacting with AWS Bedrock
type BedrockService struct {
	agentClient       *bedrockagent.Client
	agentRuntimeClient *bedrockagentruntime.Client
	region            string
	agentID           string
	agentAliasID      string
	knowledgeBaseID   string
	dataSourceID      string
}

// NewBedrockService creates a new BedrockService
func NewBedrockService() (*BedrockService, error) {
	// Get AWS region from environment
	region := os.Getenv("AWS_BEDROCK_REGION")
	if region == "" {
		return nil, fmt.Errorf("AWS_BEDROCK_REGION environment variable is not set")
	}

	// Load AWS configuration
	cfg, err := config.LoadDefaultConfig(context.Background(), config.WithRegion(region))
	if err != nil {
		return nil, fmt.Errorf("failed to load AWS config: %w", err)
	}

	// Create Bedrock agent client
	agentClient := bedrockagent.NewFromConfig(cfg)

	// Create Bedrock agent runtime client
	agentRuntimeClient := bedrockagentruntime.NewFromConfig(cfg)

	// Get required IDs from environment
	agentID := os.Getenv("AWS_BEDROCK_AGENT_ID")
	if agentID == "" {
		return nil, fmt.Errorf("AWS_BEDROCK_AGENT_ID environment variable is not set")
	}

	agentAliasID := os.Getenv("AWS_BEDROCK_AGENT_ALIAS_ID")
	if agentAliasID == "" {
		return nil, fmt.Errorf("AWS_BEDROCK_AGENT_ALIAS_ID environment variable is not set")
	}

	// Knowledge base and data source IDs are optional
	knowledgeBaseID := os.Getenv("AWS_BEDROCK_KNOWLEDGE_BASE_ID")
	dataSourceID := os.Getenv("AWS_BEDROCK_DATA_SOURCE_ID")

	return &BedrockService{
		agentClient:       agentClient,
		agentRuntimeClient: agentRuntimeClient,
		region:            region,
		agentID:           agentID,
		agentAliasID:      agentAliasID,
		knowledgeBaseID:   knowledgeBaseID,
		dataSourceID:      dataSourceID,
	}, nil
}

// FormatTraceback formats the traceback information in a Slack-friendly way
func FormatTraceback(traceback interface{}) string {
	if traceback == nil {
		return "No traceback information available"
	}

	// This is a simplified version as the Go SDK for Bedrock doesn't provide
	// the same structure as JavaScript. You may need to adjust this once you
	// have actual traceback data.
	formattedOutput := "```\n🤖 Agent Traceback\n================\n"

	// Here you would format the steps in the traceback
	// This is just a placeholder
	formattedOutput += "Traceback data not fully implemented in Go version\n"

	formattedOutput += "```"
	return formattedOutput
}

// InvokeBedrockAgent invokes the Bedrock agent with the provided input
func (s *BedrockService) InvokeBedrockAgent(inputText, sessionID string, attachments []types.FileAttachment, includeTraceback bool) (interface{}, error) {
	fmt.Printf("Session Sample ID: %s\n", sessionID)

	// Check agent health
	healthStatus, err := s.CheckBedrockAgentHealth()
	if err != nil {
		return types.ErrorResponse{
			Error:        fmt.Sprintf("Failed to check agent health: %v", err),
			OriginalError: err,
		}, nil
	}

	if !healthStatus.Healthy {
		issues := ""
		for _, issue := range healthStatus.Issues {
			issues += fmt.Sprintf("%s: %s\n", issue.Component, issue.Message)
		}
		return types.ErrorResponse{
			Error: fmt.Sprintf("AWS Bedrock agent service is not healthy: \n%s", issues),
		}, nil
	}

	// Set up the parameters for the InvokeAgent operation
	input := &bedrockagentruntime.InvokeAgentInput{
		AgentAliasId: aws.String(s.agentAliasID),
		AgentId:      aws.String(s.agentID),
		SessionId:    aws.String(sessionID),
		InputText:    aws.String(inputText),
		EnableTrace:  aws.Bool(true),
	}

	// Note: The API has changed and file attachments are no longer supported in the same way.
	// If you need file attachments, you'll need to update this code based on the new AWS SDK version.
	if len(attachments) > 0 {
		fmt.Println("Warning: File attachments are not supported in the current SDK version. Ignoring attachments.")
	}

	// Create and execute the InvokeAgent command
	output, err := s.agentRuntimeClient.InvokeAgent(context.Background(), input)
	if err != nil {
		return types.ErrorResponse{
			Error:        err.Error(),
			OriginalError: err,
		}, nil
	}

	// Get the event stream from the output
	stream := output.GetStream()
	if stream == nil {
		return "No response stream available from the agent", nil
	}

	// Read all events from the stream
	var responseText string
	var traceInfo interface{}

	// Channel to receive events
	eventsChan := stream.Events()

	// Process all events from the stream
	for event := range eventsChan {
		// Type switch to handle different event types
		switch v := event.(type) {
		case *bedrockagentruntime_types.ResponseStreamMemberChunk:
			// This is a chunk of the response text
			if v.Value.Bytes != nil && len(v.Value.Bytes) > 0 {
				// Convert bytes to string and append to response text
				responseText += string(v.Value.Bytes)
			}
		case *bedrockagentruntime_types.ResponseStreamMemberTrace:
			// This contains the trace information
			traceInfo = v.Value
		default:
			// Skip other event types (Files, ReturnControl, etc.)
			fmt.Printf("Received event of type: %T\n", v)
		}
	}

	// Check for any errors during stream processing
	if err := stream.Err(); err != nil {
		return types.ErrorResponse{
			Error:        fmt.Sprintf("Error processing stream: %v", err),
			OriginalError: err,
		}, nil
	}

	// Close the stream
	if err := stream.Close(); err != nil {
		fmt.Printf("Warning: Error closing stream: %v\n", err)
	}

	// If we didn't get any response text, use a fallback message
	if responseText == "" {
		responseText = fmt.Sprintf("Invoked agent successfully with session ID: %s, but received no response text.", sessionID)
	}

	fmt.Println("AWS Bedrock agent response:", responseText)

	// Return both response and formatted traceback if requested
	if includeTraceback {
		return types.AgentResponse{
			Response:  responseText,
			Traceback: FormatTraceback(traceInfo),
		}, nil
	}

	return responseText, nil
}

// GetKnowledgeBaseStatus gets the status of the knowledge base
func (s *BedrockService) GetKnowledgeBaseStatus() (interface{}, error) {
	if s.knowledgeBaseID == "" {
		return types.ErrorResponse{
			Error: "Knowledge base ID is not configured",
		}, nil
	}

	input := &bedrockagent.GetKnowledgeBaseInput{
		KnowledgeBaseId: aws.String(s.knowledgeBaseID),
	}

	resp, err := s.agentClient.GetKnowledgeBase(context.Background(), input)
	if err != nil {
		return types.ErrorResponse{
			Error:        err.Error(),
			OriginalError: err,
		}, nil
	}

	// Convert the response to the expected format
	return types.KnowledgeBaseStatus{
		Name:        *resp.KnowledgeBase.Name,
		Status:      string(resp.KnowledgeBase.Status),
		CreatedAt:   *resp.KnowledgeBase.CreatedAt,
		UpdatedAt:   *resp.KnowledgeBase.UpdatedAt,
		ID:          *resp.KnowledgeBase.KnowledgeBaseId,
		RawResponse: resp.KnowledgeBase,
	}, nil
}

// GetAgentStatus gets the status of the agent
func (s *BedrockService) GetAgentStatus() (interface{}, error) {
	input := &bedrockagent.GetAgentInput{
		AgentId: aws.String(s.agentID),
	}

	resp, err := s.agentClient.GetAgent(context.Background(), input)
	if err != nil {
		return types.ErrorResponse{
			Error:        err.Error(),
			OriginalError: err,
		}, nil
	}

	// Convert the response to the expected format
	foundationModel := ""
	if resp.Agent.FoundationModel != nil {
		foundationModel = *resp.Agent.FoundationModel
	}

	return types.AgentStatus{
		AgentName:       *resp.Agent.AgentName,
		AgentID:         *resp.Agent.AgentId,
		AgentStatus:     string(resp.Agent.AgentStatus),
		FoundationModel: foundationModel,
		CreatedAt:       *resp.Agent.CreatedAt,
		UpdatedAt:       *resp.Agent.UpdatedAt,
		RawResponse:     resp.Agent,
	}, nil
}

// GetDataSource gets information about the data source
func (s *BedrockService) GetDataSource() (interface{}, error) {
	if s.knowledgeBaseID == "" || s.dataSourceID == "" {
		return types.ErrorResponse{
			Error: "Knowledge base ID or data source ID is not configured",
		}, nil
	}

	// Retrieve the last ingestion job for the data source
	input := &bedrockagent.ListIngestionJobsInput{
		KnowledgeBaseId: aws.String(s.knowledgeBaseID),
		DataSourceId:    aws.String(s.dataSourceID),
		MaxResults:      aws.Int32(1),
		// SortBy is not available in the current version, so we're not setting it
	}

	resp, err := s.agentClient.ListIngestionJobs(context.Background(), input)
	if err != nil {
		return types.ErrorResponse{
			Error:        err.Error(),
			OriginalError: err,
		}, nil
	}

	// Check if we have ingestion jobs
	if resp.IngestionJobSummaries == nil || len(resp.IngestionJobSummaries) == 0 {
		return struct {
			Status  string `json:"status"`
			Message string `json:"message"`
		}{
			Status:  "NO_JOBS_FOUND",
			Message: "No data sources found for this knowledge base.",
		}, nil
	}

	// Return information about the most recent job
	job := resp.IngestionJobSummaries[0]
	return types.DataSourceInfo{
		DataSourceID:    *job.DataSourceId,
		KnowledgeBaseID: *job.KnowledgeBaseId,
		Description:     *job.Description,
		Status:          string(job.Status),
		StartedAt:       *job.StartedAt,
		UpdatedAt:       *job.UpdatedAt,
		RawResponse:     job,
	}, nil
}

// SyncDataSource triggers a synchronization of the data source
func (s *BedrockService) SyncDataSource() (interface{}, error) {
	if s.knowledgeBaseID == "" || s.dataSourceID == "" {
		return types.ErrorResponse{
			Error: "Knowledge base ID or data source ID is not configured",
		}, nil
	}

	input := &bedrockagent.StartIngestionJobInput{
		KnowledgeBaseId: aws.String(s.knowledgeBaseID),
		DataSourceId:    aws.String(s.dataSourceID),
		Description:     aws.String("Manual sync triggered on " + time.Now().Format(time.RFC3339)),
	}

	resp, err := s.agentClient.StartIngestionJob(context.Background(), input)
	if err != nil {
		return types.ErrorResponse{
			Error:        err.Error(),
			OriginalError: err,
		}, nil
	}

	return types.DataSourceSync{
		DataSourceID:    *resp.IngestionJob.DataSourceId,
		KnowledgeBaseID: *resp.IngestionJob.KnowledgeBaseId,
		IngestionJobID:  *resp.IngestionJob.IngestionJobId,
		Status:          string(resp.IngestionJob.Status),
		RawResponse:     resp.IngestionJob,
	}, nil
}

// GetDataSourceConfig gets the configuration of the data source
func (s *BedrockService) GetDataSourceConfig() (interface{}, error) {
	if s.knowledgeBaseID == "" || s.dataSourceID == "" {
		return types.ErrorResponse{
			Error: "Knowledge base ID or data source ID is not configured",
		}, nil
	}

	input := &bedrockagent.GetDataSourceInput{
		KnowledgeBaseId: aws.String(s.knowledgeBaseID),
		DataSourceId:    aws.String(s.dataSourceID),
	}

	resp, err := s.agentClient.GetDataSource(context.Background(), input)
	if err != nil {
		return types.ErrorResponse{
			Error:        err.Error(),
			OriginalError: err,
		}, nil
	}

	return types.DataSourceConfig{
		Name:              *resp.DataSource.Name,
		Status:            string(resp.DataSource.Status),
		ConfigurationType: string(resp.DataSource.DataSourceConfiguration.Type),
		CreatedAt:         *resp.DataSource.CreatedAt,
		UpdatedAt:         *resp.DataSource.UpdatedAt,
		RawResponse:       resp.DataSource,
	}, nil
}

// ListDataSources lists all data sources
func (s *BedrockService) ListDataSources() (interface{}, error) {
	if s.knowledgeBaseID == "" {
		return types.ErrorResponse{
			Error: "Knowledge base ID is not configured",
		}, nil
	}

	input := &bedrockagent.ListDataSourcesInput{
		KnowledgeBaseId: aws.String(s.knowledgeBaseID),
	}

	resp, err := s.agentClient.ListDataSources(context.Background(), input)
	if err != nil {
		return types.ErrorResponse{
			Error:        err.Error(),
			OriginalError: err,
		}, nil
	}

	dataSources := []types.DataSource{}
	for _, source := range resp.DataSourceSummaries {
		dataSources = append(dataSources, types.DataSource{
			DataSourceID: *source.DataSourceId,
			Name:         *source.Name,
			Status:       string(source.Status),
			UpdatedAt:    *source.UpdatedAt,
			RawResponse:  source,
		})
	}

	return types.DataSourceList{
		DataSources: dataSources,
		Count:       len(dataSources),
	}, nil
}

// GetIngestionJobStatus gets the status of an ingestion job
func (s *BedrockService) GetIngestionJobStatus(jobID string) (interface{}, error) {
	if s.knowledgeBaseID == "" || s.dataSourceID == "" {
		return types.ErrorResponse{
			Error: "Knowledge base ID or data source ID is not configured",
		}, nil
	}

	input := &bedrockagent.GetIngestionJobInput{
		KnowledgeBaseId: aws.String(s.knowledgeBaseID),
		DataSourceId:    aws.String(s.dataSourceID),
		IngestionJobId:  aws.String(jobID),
	}

	resp, err := s.agentClient.GetIngestionJob(context.Background(), input)
	if err != nil {
		return types.ErrorResponse{
			Error:        err.Error(),
			OriginalError: err,
		}, nil
	}

	failureReasons := []string{}
	for _, reason := range resp.IngestionJob.FailureReasons {
		failureReasons = append(failureReasons, reason)
	}

	statistics := ""
	if resp.IngestionJob.Statistics != nil {
		// Format statistics based on what's available in the current SDK version
		statistics = fmt.Sprintf("Status: %s", resp.IngestionJob.Status)
	}

	return types.IngestionJobStatus{
		IngestionJobID:  *resp.IngestionJob.IngestionJobId,
		Status:          string(resp.IngestionJob.Status),
		StartedAt:       *resp.IngestionJob.StartedAt,
		UpdatedAt:       *resp.IngestionJob.UpdatedAt,
		Statistics:      statistics,
		FailureReasons:  failureReasons,
		RawResponse:     resp.IngestionJob,
	}, nil
}

// CheckBedrockAgentHealth checks the health of the Bedrock agent
func (s *BedrockService) CheckBedrockAgentHealth() (types.HealthStatus, error) {
	issues := []types.HealthIssue{}
	details := types.HealthDetails{
		Region:       s.region,
		AgentID:      s.agentID,
		AgentAliasID: s.agentAliasID,
	}

	// Check agent status
	agentStatusResp, err := s.GetAgentStatus()
	if err != nil {
		return types.HealthStatus{}, err
	}

	// Check if response contains an error
	if errorResp, ok := agentStatusResp.(types.ErrorResponse); ok {
		issues = append(issues, types.HealthIssue{
			Component: "Agent",
			Status:    "ERROR",
			Message:   fmt.Sprintf("Failed to check Agent status: %s", errorResp.Error),
		})
	} else if agentStatus, ok := agentStatusResp.(types.AgentStatus); ok {
		details.AgentName = agentStatus.AgentName

		if agentStatus.AgentStatus != "PREPARED" && agentStatus.AgentStatus != "READY" {
			issues = append(issues, types.HealthIssue{
				Component: "Agent",
				Status:    agentStatus.AgentStatus,
				Message:   fmt.Sprintf("Agent is not in a ready state. Current status: %s", agentStatus.AgentStatus),
			})
		}
	}

	// Check knowledge base status if configured
	if s.knowledgeBaseID != "" {
		kbStatusResp, err := s.GetKnowledgeBaseStatus()
		if err != nil {
			return types.HealthStatus{}, err
		}

		// Check if response contains an error
		if errorResp, ok := kbStatusResp.(types.ErrorResponse); ok {
			issues = append(issues, types.HealthIssue{
				Component: "Knowledge Base",
				Status:    "ERROR",
				Message:   fmt.Sprintf("Failed to check Knowledge Base status: %s", errorResp.Error),
			})
		} else if kbStatus, ok := kbStatusResp.(types.KnowledgeBaseStatus); ok {
			details.KnowledgeBaseName = kbStatus.Name

			if kbStatus.Status != "ACTIVE" {
				issues = append(issues, types.HealthIssue{
					Component: "Knowledge Base",
					Status:    kbStatus.Status,
					Message:   fmt.Sprintf("Knowledge Base is not active. Current status: %s", kbStatus.Status),
				})
			}
		}
	}

	// Return final health status
	return types.HealthStatus{
		Healthy: len(issues) == 0,
		Issues:  issues,
		Details: details,
	}, nil
}

// MonitorIngestionJob monitors an ingestion job
func (s *BedrockService) MonitorIngestionJob(jobID string, maxWaitMinutes int) (interface{}, error) {
	// Store initial KB timestamp
	kbStatusResp, err := s.GetKnowledgeBaseStatus()
	if err != nil {
		return types.ErrorResponse{
			Error:        fmt.Sprintf("Failed to get initial knowledge base status: %v", err),
			OriginalError: err,
		}, nil
	}

	var initialKBTimestamp time.Time
	if kbStatus, ok := kbStatusResp.(types.KnowledgeBaseStatus); ok {
		initialKBTimestamp = kbStatus.UpdatedAt
	} else {
		return types.ErrorResponse{
			Error: "Failed to get initial knowledge base timestamp",
		}, nil
	}

	// Monitor the job
	jobComplete := false
	var jobStatusString string

	// Set timeout
	timeout := time.Now().Add(time.Duration(maxWaitMinutes) * time.Minute)

	// Poll every 30 seconds
	for time.Now().Before(timeout) && !jobComplete {
		jobStatusResp, err := s.GetIngestionJobStatus(jobID)
		if err != nil {
			return types.ErrorResponse{
				Error:        fmt.Sprintf("Failed to get job status: %v", err),
				OriginalError: err,
			}, nil
		}

		// Check if response contains an error
		if errorResp, ok := jobStatusResp.(types.ErrorResponse); ok {
			return errorResp, nil
		}

		if jobStatus, ok := jobStatusResp.(types.IngestionJobStatus); ok {
			jobStatusString = jobStatus.Status

			// Check if job is complete
			if jobStatus.Status == "COMPLETE" || jobStatus.Status == "FAILED" || jobStatus.Status == "STOPPED" {
				jobComplete = true
				break
			}
		}

		// Wait before checking again
		time.Sleep(30 * time.Second)
	}

	if !jobComplete {
		return types.ErrorResponse{
			Error: fmt.Sprintf("Job monitoring timed out after %d minutes", maxWaitMinutes),
		}, nil
	}

	// Get final KB timestamp
	kbStatusResp, err = s.GetKnowledgeBaseStatus()
	if err != nil {
		return types.ErrorResponse{
			Error:        fmt.Sprintf("Failed to get final knowledge base status: %v", err),
			OriginalError: err,
		}, nil
	}

	var finalKBTimestamp time.Time
	var kbID string
	if kbStatus, ok := kbStatusResp.(types.KnowledgeBaseStatus); ok {
		finalKBTimestamp = kbStatus.UpdatedAt
		kbID = kbStatus.ID
	} else {
		return types.ErrorResponse{
			Error: "Failed to get final knowledge base timestamp",
		}, nil
	}

	// Check if KB was updated
	kbUpdated := finalKBTimestamp.After(initialKBTimestamp)

	// Get agent status to ensure it's ready
	agentStatusResp, err := s.GetAgentStatus()
	if err != nil {
		return types.ErrorResponse{
			Error:        fmt.Sprintf("Failed to get agent status: %v", err),
			OriginalError: err,
		}, nil
	}

	var agentReady bool
	var agentStatus types.AgentStatus
	if agentStatusVal, ok := agentStatusResp.(types.AgentStatus); ok {
		agentStatus = agentStatusVal
		agentReady = agentStatusVal.AgentStatus == "READY" || agentStatusVal.AgentStatus == "PREPARED"
	} else {
		return types.ErrorResponse{
			Error: "Failed to get agent status",
		}, nil
	}

	// Determine success message
	message := ""
	if jobStatusString == "COMPLETE" && kbUpdated && agentReady {
		message = "Knowledge base successfully updated and agent is ready"
	} else if jobStatusString == "COMPLETE" && !kbUpdated {
		message = "Job completed but knowledge base timestamp was not updated"
	} else if jobStatusString == "COMPLETE" && !agentReady {
		message = "Job completed but agent is not in a ready state"
	} else {
		message = fmt.Sprintf("Job status: %s", jobStatusString)
	}

	// Return monitoring results
	return types.MonitorIngestionJobStatus{
		Success:            jobStatusString == "COMPLETE" && kbUpdated && agentReady,
		KnowledgeBaseID:    kbID,
		DataSourceID:       s.dataSourceID,
		IngestionJobID:     jobID,
		JobStatus:          jobStatusString,
		InitialKBTimestamp: initialKBTimestamp,
		FinalKBTimestamp:   finalKBTimestamp,
		KBUpdated:          kbUpdated,
		AgentStatus:        agentStatus,
		AgentReady:         agentReady,
		Message:            message,
	}, nil
}
